driver/src/orderbook/streamed/orderbook.rs

```diff
@@ -0,0 +1,84 @@
+use super::*;
+use crate::{
+    contracts::stablex_contract::batch_exchange,
+    models::{AccountState, Order},
+    orderbook::StableXOrderBookReading,
+};
+use anyhow::Result;
+use ethcontract::{contract::EventData, H256, U256};
+use state::{Batch, State};
+use std::collections::BTreeMap;
+
+// Ethereum events (logs) can be both created and removed. Removals happen if the chain reorganizes
+// and ends up not including block that was previously thought to be part of the chain.
+// However, the orderbook state (`State`) cannot remove events. To support this, we keep an ordered
+// list of all events based on which the state is built.
+
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd)]
+struct EventSortKey {
+    block_number: u64,
+    /// Is included to differentiate events from the same block number but different blocks which
+    /// can happen during reorgs.
+    block_hash: H256,
```

Comment:
Should `log_index` come before `block_hash` in terms of ordering?

Reply:
It does not matter. The only time during which there can be multiple blocks of the same number is during reorg before all old events have been deleted. And in that case the resulting state isn't correct anyway.

Reply:
Actually in that case, having `block_hash` come first as it is makes a little more sense, as the events will be grouped together.